### PR TITLE
feat(tables): score competing line-table candidates

### PR DESCRIPTION
**Pull request by abimaelmartell**
## Summary

Compare legacy ruled-table output with strict dense-row and open-edge grid hypotheses, then select the candidate with the strongest occupied-cell and source-item evidence. This recovers multi-level booktabs and open-sided tables without accepting separated chart bands or sparse decorative rules.

OpenDataLoader improves from 0.872383 to 0.873213 overall, including TEDS rising from 0.795689 to 0.814117.

---
## Summary by cubic
Adds evidence-based scoring to pick the best line-table among the legacy grid and two new hypotheses. Recovers multi-level booktabs and open-edge tables (including two-column grids), preserves multi-line open-edge headers, and retains multiple open-edge bands on one page; accuracy improves (OpenDataLoader 0.872383→0.873213; TEDS 0.795689→0.814117).

- **New Features**
  - Dense-row anchor hypothesis: builds columns from repeated dense rows with numeric body evidence.
  - Open-edge grid hypothesis: uses horizontal outer bounds + interior dividers; pulls header above the top rule.
  - Evidence scoring and selection: compares legacy vs alternatives, filters overlaps, and keeps winners in `detect_tables_from_lines`.

- **Bug Fixes**
  - Stronger validation: dense-row requires a contiguous band with ≥2 spanning rules and rejects uniform rule runs; open-edge requires tall interior dividers, occupied body rows, and all columns occupied.
  - Prevent synthetic merges and keep multiple open-edge grids on the same page; drop alternatives that span multiple sparse-anchor tables and re-evaluate remaining geometry without re-running anchors.
  - Recover two-column open-edge grids with a single interior divider; allow an unlabeled first stub while requiring other header columns to be populated.

<sup>Written for commit 35e244fde9add5493642e2ac4c7228c61b04ced2. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/firecrawl/pdf-inspector/pull/177?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by abimaelmartell** [comment]
@cubic-dev-ai Please review exact head 48cdc60. This PR compares legacy ruled-table output with dense-row and open-edge candidates using deterministic evidence scoring. Full fmt/clippy/test/release gates pass (694 unit / 141 integration), OpenDataLoader improves 0.872383 -> 0.873213 and remains +0.003487 over LiteParse with zero document regressions, and pdf-evals remains 202 passed / 1 pre-existing unsnapshotted extraction.

**Comment by abimaelmartell** [comment]
@cubic-dev-ai Please start a fresh review for exact head 48cdc60. The existing Cubic suite has remained queued with zero runs; the PR was reopened without any commit or tree change, and all CI/benchmark/regression gates remain green.

**Comment by abimaelmartell** [comment]
@cubic-dev-ai Please review exact head 04324bf. This is a metadata-only SHA refresh of the already validated tree, pushed while the PR is open and ready so the review app receives a clean synchronize event. All CI, benchmark, and regression evidence remains unchanged.

**Comment by abimaelmartell** [comment]
Addressed the Cubic findings on exact head `37c3670`:

- made candidate scoring saturating for sparse/empty tables
- allowed populated first-column headers for coherent open-edge grids
- moved dense/open-edge alternatives ahead of sparse-anchor early returns
- scoped dense-row vertical-line vetoes to the candidate band
- count logical fragments and distinct matched anchor columns
- select non-overlapping competing hypotheses while rejecting page-wide merges of independent sparse tables
- kept populated-header inference behind a single-span ambiguity guard and prevented recursive sparse re-detection

Validation: `cargo fmt`, `cargo clippy -- -D warnings`, `cargo test` (701 unit + 141 integration + 2 doc), release build, OpenDataLoader 0.873213 (+0.003487 vs LiteParse; 2 improved / 0 regressed), and pdf-evals 202 passed / 1 pre-existing unsnapshotted fixture.

@cubic-dev-ai please re-review this exact head.

File: src/tables/detect_lines.rs [hunk 1]
```diff
@@ -671,6 +671,441 @@ fn combine_non_overlapping_tables(mut primary: Vec<Table>, secondary: Vec<Table>
...
+    if occupied_body_rows < 2 || occupied_body_columns != column_count {
+        return None;
+    }
+
+    let header_band = [
+        (y_top + 30.0, x_min, x_max),
+        (y_top + RULE_Y_TOLERANCE, x_min, x_max),
+    ];
+    let header_rows = collect_anchored_rows(items, &header_band, page);
+    let (_, header_items) = header_rows.last()?;
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/tables/detect_lines.rs]: Fixed in 228a3bf by merging every header baseline into its assigned column in top-to-bottom order, preserving all item indices. Added a two-line, three-column regression test; full Rust, OpenDataLoader, and pdf-evals gates remain green.

**Comment by abimaelmartell** [comment]
Addressed the remaining multi-line open-edge header finding on exact head `228a3bf`: all header baselines are now merged per column in reading order, with all source items retained. Added a focused two-line header regression test.

Validation remains green: 702 unit + 141 integration + 2 doc tests, clippy/format/release build, OpenDataLoader 0.873213 (+0.003487 vs LiteParse; 2 improved / 0 regressed), and pdf-evals 202 passed / 1 pre-existing unsnapshotted fixture.

@cubic-dev-ai please re-review this exact head.

File: src/tables/detect_lines.rs [hunk 2]
```diff
@@ -671,6 +671,443 @@ fn combine_non_overlapping_tables(mut primary: Vec<Table>, secondary: Vec<Table>
...
+            *x > x_min + RULE_JOIN_GAP
+                && *x < x_max - RULE_JOIN_GAP
+                && *y_min <= y_bottom + RULE_Y_TOLERANCE
+                && *y_max >= y_top - RULE_Y_TOLERANCE
+                && *y_max - *y_min >= height * 0.8
+        })
+        .map(|(x, _, _)| *x)
+        .collect();
+    let interior_edges = snap_edges(&scoped_vertical_xs, 3.0);
+    if !(2..=24).contains(&interior_edges.len()) {
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/tables/detect_lines.rs]: Fixed in 3327ccd: two-column open-edge grids now accept their single interior divider. The remaining-page pass also keeps a geometry-only fallback so an inferred two-column header cannot replace an adjacent physical grid; the new focused regression test and the existing sparse/vector coexistence test both pass.

**Comment by abimaelmartell** [comment]
@cubic-dev-ai Please re-review exact head 3327ccd. This addresses the two-column open-edge case and preserves independent physical grids when inferred headers overlap sparse anchor bands. Full Rust gates, OpenDataLoader (0.873213; 0 regressed), and pdf-evals (202 pass / 1 pre-existing unsnapshotted) are green.

File: src/tables/detect_lines.rs [hunk 3]
```diff
@@ -671,6 +671,443 @@ fn combine_non_overlapping_tables(mut primary: Vec<Table>, secondary: Vec<Table>
...
+    }
+    if rows.len() > distinct_rule_ys.len() * 2 + 2 {
+        // Sparse page decorations around prose can expose many aligned text
+        // starts, but their rule levels do not corroborate the row schema.
+        return None;
+    }
+    let anchors = rows
+        .iter()
+        .map(|(_, row)| logical_row_anchors(row))
+        .max_by_key(Vec::len)?;
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/tables/detect_lines.rs:819]: Fixed in 35e244f: open-edge detection now builds a candidate for every qualifying span/run, scores all non-overlapping hypotheses together, and scopes the populated-header ambiguity guard to the local ruled band. The new end-to-end test preserves two differently sized open-edge tables on one page.

**Comment by abimaelmartell** [comment]
@cubic-dev-ai Please re-review exact head 35e244f. The multi-band open-edge issue is fixed with per-span/per-run candidates and non-overlapping hypothesis selection. Full Rust gates, OpenDataLoader (0.873213; 0 regressed), and pdf-evals (202 pass / 1 pre-existing unsnapshotted) are green.
